postmortems/001 - Postmortem for v0.0.19 Deployment Failure.md

```diff
@@ -0,0 +1,54 @@
+# 000 - Postmortem for v0.0.19 Deployment Failure
+
+## Level: Internal
+
+## Author(s)
+
+- David Ellis <david@iasql.com>
+
+## Summary
+
+The `v0.0.19` release could not be deployed despite the test suite succeeding. A combination of a new deployment mechanism and gaps in the test suite actually *prevented* a worse outage of deploying broken code, but feature development was significantly slowed trying to figure out what exactly broke. It turned out to be a mistake in the `iasql_upgrade` logic, made more difficult to figure out due to separate-but-unrelated bugs with the new `default_aws_region` postgres function. The new deployment mechanism exercises `iasql_upgrade` and therefore hit this bug and prevented rollout of the faulty release.
+
+## Timeline
+
+- **2022-09-05**: Release day for `v0.0.19`, rushing to get desired features out the door, but were unable to. Release was delayed rather than skip these features.
+- **2022-09-07**: [Actual attempted release for `v0.0.19`](https://github.com/iasql/iasql-engine/tree/v0.0.19), tests are passing on the `main` branch, and manual testing locally did not uncover any issues. There *was* an error in the staging deploy earlier in the day, but it was assumed caused by other experiments. After the `staging` IaSQL database was cleaned up, further staging deploys were successful, so it was assumed that the `production` deployment would be the same. Instead the `production` deploy failed, and an investigation began.
+- **2022-09-09**: [Actual root cause found and fixed, and new test written to verify no regressions occur](https://github.com/iasql/iasql-engine/pull/1241). A few bugs unrelated to the root cause in the `default_aws_region` function are also fixed in the process.
+- **2022-09-12**: Release day for `v0.0.20`, but being extra cautious and it was also put on hold because of an error in the Prisma Example end-to-end test, and investigation into this occurs.
+- **2022-09-13**: [Improved logging in the Prisma example deployment](https://github.com/iasql/iasql-engine/pull/1270), but cause of this issue not found.
+- **2022-09-14**: [Root cause of Prisma end-to-end test found](https://github.com/iasql/iasql-engine/pull/1272): The test account was not properly cleaned. Adding a pre-clean script however, also broke things, and an issue in the `delete_all_records` postgres function was discovered. Adding `aws_regions` to a blacklist to that function resolved that issue. Deployment of `v0.0.20` successfully occurred after that and the Prisma end-to-end test fix was not truly necessary as `delete_all_records` is not meant for general consumption (yet).
```

Comment:
Were these 2 events directly related to the `0.0.19` issue? I understand they appeared and slow a bit `0.0.20`, but not sure if they are part of this postmortem.

Reply:
I included them in the timeline because it was work that was done before the outage was resolved and was done as part of resolving the outage.

It was only realized afterwards that it was not related to the bug that prevented `v0.0.19` from being deployed.